### PR TITLE
Add a client object to libhoney for isolating traffic

### PR DESCRIPTION
**Pull request by maplebed**
This change adds a `Client` object to libhoney. 

When sending traffic of different importance, it can be frustrating to have them run through the same transmission queue. It would be nice to be able to identify some traffic as important enough to block and slow down if sending backs up while allowing other (say logging) traffic to be dropped. With one transmission and a global `BlockOnSend` that's not possible. This change introduces a libhoney Client with its own queues and transmission. It does not share any state with the global client.

Addresses #11, not by removing the global state, but by giving libhoney users a method of ignoring it.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,146 @@
+package libhoney
+
+import (
+	"errors"
+)
+
+// TODO think about whether it would be useful for Client to be an interface instead
```
Comment (tredman) [inline client.go]: looks like an interface now :)

File: libhoney.go [hunk 2]
```diff
@@ -247,6 +251,9 @@ type Builder struct {
 	// any dynamic fields to apply to each generated event
 	dynFields     []dynamicField
 	dynFieldsLock sync.RWMutex
+
+	// client is the Client to use to send events generated from this builder
+	client *defaultClient
```
Comment (tredman) [inline libhoney.go]: Should this be a `Client` (interface) rather than a specific struct?

File: libhoney.go [hunk 3]
```diff
@@ -203,6 +204,9 @@ type Event struct {
 
 	// fieldHolder contains fields (and methods) common to both events and builders
 	fieldHolder
+
+	// client is the Client to use to send events generated from this builder
+	client *defaultClient
```
Comment (tredman) [inline libhoney.go]: same here - `Client` rather than `defaultClient`

File: libhoney.go [hunk 4]
```diff
@@ -428,41 +445,87 @@ func SendNow(data interface{}) error {
...
+
+// NewBuilder creates a new event builder. The builder inherits any
+// Dynamic or Static Fields present in the global scope.
+func NewBuilder() *Builder {
+	ensureDefaultBuilder()
+	startTx()
+	return dc.defaultBuilder.Clone()
+}
+
+func ensureDefaultBuilder() {
```
Comment (tredman) [inline libhoney.go]: Is this lazy initialization to prevent the application from experiencing errors if they used libhoney, but haven't called `Init()`? In the python sdk we lean towards just no-oping and printing a warning if uninitialized. The Tx/Builder here wouldn't work due to lack of a valid API key if Init hasn't been called.

Comment (maplebed) [reply libhoney.go]: You can choose to explicitly set the API key etc. on a per-builder or per-event basis, so there is a path to skipping initialization. I suppose noop is an alternative, though there are a lot of nil references to protect if it's uninitialized.

**Comment by maplebed** [comment]
ok, making Client an interface didn't work at all. I've changed it so that Transmission is now off in its own package, Responses are owned by Transmissions, and if you want mocks, you mock Transmission, not Client.

File: client.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		fieldHolder: fieldHolder{
+			data: make(map[string]interface{}),
+		},
+		client: c,
+	}
+
+	return c, nil
+}
+
+func (c *Client) ensureTransmission() {
```
Comment (ianwilkes) [inline client.go]: All of these `ensure` methods are super not concurrency safe. Add a test which tries `Add()` etc on a shared, empty client from lots of goroutines and see what `go test -race` has to say about it. You're gonna need some locking if these fields are allowed to be set after `NewClient`.

However, for `logger` for example, you could instead add a `logf` method to `Client`, have everything call that instead of `c.logger.Printf`, and let that method handle the default case. You could include a `NullLogger` instance in Client for such cases even if you're not always using it.

`transmission` is harder but you could add methods for that too. Maybe. I'm not happy about the fact that if a method is called with a `nil` receiver, you end up creating and starting a new Transmission in each method call. It's such an edge case, maybe just blackhole the data instead... or don't handle the nil. A panic in a library is acceptable in the event of developer error.

File: client.go [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+// them somewhere. It maintains its own sending queue for events, distinct from
+// both the package-level libhoney queue and any other client. Clients should be
+// created with NewClient(config). A manually created Client{} will function as
+// a nil output and drop everything handed to it (so can be safely used in dev
+// and tests). For more complete testing you can create a Client with a
+// MockOutput transmission then inspect the events it would have sent.
+type Client struct {
+	transmission transmission.Sender
+	logger       Logger
+	builder      *Builder
```
Comment (ianwilkes) [inline client.go]: Is a `Builder` something that's valid even when empty? If so, you could just have type `Builder` here and then you'd never confront the `builder == nil` problem.

File: transmission/writer.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+func (w *WriterOutput) Start() error {
+	w.responses = make(chan Response, w.ResponseQueueSize)
+	return nil
+}
+
+func (w *WriterOutput) Stop() error { return nil }
+
+func (w *WriterOutput) Add(ev *Event) {
+	var m []byte
+	func() {
```
[the remark this answers is not in the document]
Comment (maplebed) [reply transmission/writer.go]: wtf I have no idea why.

File: transmission/writer.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+		m = append(m, '\n')
+	}()
+
+	w.Lock()
+	defer w.Unlock()
+	if w.W == nil {
+		w.W = os.Stdout
+	}
+	w.W.Write(m)
+	// TODO add response to the response queue
```
[the remark this answers is not in the document]
Comment (maplebed) [reply transmission/writer.go]: In this push of code (already fixed, but not checked in) the Honeycomb Sender is the only one that actually pushes anything on to the Response queue. If the expected API of a Sender is that every event that gets `Add`ed generates a response on the response queue, none of the Senders fulfill that promise.  In the pre-client world, it was actually impossible for any implementation of `Output` (besides the default Honeycomb one) to push anything on to the Responses queue.  It is now possible, and every implementation of the Sender interface should do so. That's what that comment is saying to do.

File: transmission/writer.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (maplebed) [reply transmission/writer.go]: ah. In the old version, there was a lock that was taken at the beginning of the closure and a `defer`ed unlock. That lock is now gone, though the closure remained. It can now go too.

**Comment by maplebed** [comment]
Re-requested reviews. Let's see how we do for round 3

File: client.go (hunk 6 above)
Comment (maplebed) [reply client.go]: I put some `sync.Once`es in the `ensure`s. An empty builder isn't really a thing I want to start mucking with; they should be created by cloning an existing builder or using `NewBuilder()`.

File: client.go (hunk 5 above)
Comment (maplebed) [reply client.go]: I sprinkled some `sync.Once` all over it. Tastes better, less filling.

File: libhoney.go [hunk 9]
```diff
@@ -636,42 +723,25 @@ func (e *Event) SendPresampled() (err error) {
...
-	return nil
-}
-
-// sendResponse sends a dropped event response down the response channel
-func sendDroppedResponse(e *Event, message string) {
-	r := Response{
-		Err:      errors.New(message),
-		Metadata: e.Metadata,
+	e.client.ensureTransmission()
+	txEvent := &transmission.Event{
```
Comment (tredman) [inline libhoney.go]: seems unfortunate to have to duplicate the event struct and copy everything over. Assuming this was to prevent circular imports? Could we eventually refactor event into its own package?

File: writer.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
+package libhoney
+
+import (
+	"github.com/honeycombio/libhoney-go/transmission"
+)
+
+// WriterOutput implements the Output interface and passes it along to the
+// transmission.WriterSender. It is deprecated and you sholud use the
```
Comment (tredman) [inline writer.go]: nit: `should`

File: libhoney.go (hunk 9 above)
Comment (maplebed) [reply libhoney.go]: We could, but not easily. The public interface of the SDK exports `libhoney.Event` as a thing, so I need to maintain that. We could move Event to its own package and then have libhoney.Event include that type (in the same way there's a `transitionalOutput` type), but it gets messy with the fieldholder. This diff is big enough; I'd rather keep it as is for now and try and move it later if it remains an issue.

File: transmission/writer.go [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+	}
+	w.responses = make(chan Response, w.ResponseQueueSize)
+	return nil
+}
+
+func (w *WriterSender) Stop() error { return nil }
+
+func (w *WriterSender) Add(ev *Event) {
+	var m []byte
+		tPointer := &(ev.Timestamp)
```
[the remark this answers is not in the document]
Comment (maplebed) [reply transmission/writer.go]: yup. fixed

File: client.go [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+		}
+	} else {
+		c.transmission = conf.Transmission
+	}
+	if err := c.transmission.Start(); err != nil {
+		c.logger.Printf("transmission client failed to start: %s", err.Error())
+		return nil, err
+	}
+
+	c.builder = &Builder{
```
[the remark this answers is not in the document]
Comment (maplebed) [reply client.go]: Nope - you can't get a new client from a client, only builders.

**Comment by maplebed** [comment]
> in general I'm not the biggest fan of APIs that permit both zero-initialized allocation forms (`&Client`) as well as a constructor+config options (`NewClient`). _So_ many concurrency problems go away if we only support the latter and guarantee that the instance is fully initialized before being returned. Why do we need both again?

The usual way to explicitly prohibit empty initialization of the struct is to make the Client an interface and the default implementation an unexported struct but doing that in this case caused so many more problems than it solves... 

The pattern of "an empty struct will noop; use the constructor for one that works" seems like a reasonable alternative to me. In this case, if you make a `&Client{}`, you'll get one with a `DiscardSender` as the transmission and all your instrumentation will go in to a black hole. Which might actually be a feature - for all your tests you don't need to construct a working discard client, you can just use an uninitialized one.

File: client.go [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+				client: c,
+			}
+		}
+	})
+}
+
+// Close waits for all in-flight messages to be sent. You should
+// call Close() before app termination.
+func (c *Client) Close() {
+	if c == nil {
```
[the remark this answers is not in the document]
Comment (maplebed) [reply client.go]: except that by function of an empty Client{} being a working (discard) client, creating the thing then throwing it out means that a nil `c.NewBulider().NewEvent()` would still create a valid event. 
but ...
I am also on the fence and am comfortable saying that this should panic. It's a tall order - IMO an instrumentation library should _never_ panic - blowing up the service because the instrumentation is broken feels extraordinarily rude. But I also agree that `c` being `nil` is 99% likely an error (and one that should be found, because the absence of instrumentation may incorrectly be interpreted as the instrumented thing having not run), so we should help find the error via panic.

I'm going to remove the nil receiver checks.

File: libhoney.go [hunk 14]
```diff
@@ -48,24 +51,21 @@ var (
...
-	defaultBuilder   = &Builder{
-		APIHost:    defaultAPIHost,
-		SampleRate: defaultSampleRate,
-		dynFields:  make([]dynamicField, 0, 0),
-		fieldHolder: fieldHolder{
-			data: make(map[string]interface{}),
-		},
-	}
+	// singleton-like client used if you use package-level functions
+	dc = &Client{}
```
[the remark this answers is not in the document]
Comment (maplebed) [reply libhoney.go]: I'm not super happy with using constructors in global variables. Given that I do want an empty `&Client{}` to be a valid thing, this doesn't seem like a bad choice. The ensures will have to be there anyways (and they're relatively light, and they're only on package-level functions, which are called rarely in the normal use of things).

**Comment by toshok** [comment]
> The pattern of "an empty struct will noop; use the constructor for one that works" seems like a reasonable alternative to me

Sure, but it doesn't noop - it doesn't do any outwardly visible work, but it still _does_ stuff, and is the reason for most of the concurrency cognitive load, right?  If we never had to worry about lazy initializing an uninitialized `&Client{}`, we wouldn't need the `sync.Once`'s, we wouldn't need tests for data races, etc.

If the uninitialized path is _only_ meant for a noop client, why not a private `initialized` field in `Client` that is only set to `true` in the constructor.  Then we can get rid of (i think?) all the `ensure*` calls because we can guarantee that a client is either fully initialized or not initialized at all.  the nil client checks could early return on it too, `if client == nil || !client.initialized`

**Comment by toshok** [comment]
Like I said in slack, feels like I came in a little late to this PR to give these sorts of comments.  It's also not going to be an outwardly visible change, so if what's there works, 👍 

From a maintenance perspective, though, I can't think of a good argument for continuing with the all the concurrency stuff, though.  future PR?

File: client.go [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+		conf.APIHost = defaultAPIHost
+	}
+	if conf.Dataset == "" {
+		conf.Dataset = defaultDataset
+	}
+
+	c := &Client{
+		logger: conf.Logger,
+	}
+	c.ensureLogger()
```
[the remark this answers is not in the document]
Comment (maplebed) [reply client.go]: I’m not clear why that would be better. The way it is not the sync.Once gets triggered during initialization so all future calls quickly noop. Delaying that till the first place they’re hit makes for a more likely collision (and therefore block) doesn’t it? 

Can you say more about the motivation for this suggested change?

File: libhoney.go [hunk 16]
```diff
@@ -313,91 +450,14 @@ type dynamicField struct {
...
-	return nil
-}
-
 // Close waits for all in-flight messages to be sent. You should
 // call Close() before app termination.
 func Close() {
-	logger.Printf("closing libhoney")
-	if tx != nil {
-		tx.Stop()
+	dc.ensureLogger()
```
[the remark this answers is not in the document]
Comment (maplebed) [reply libhoney.go]: yup! 👍

File: libhoney.go [hunk 17]
```diff
@@ -580,6 +668,57 @@ func (f *fieldHolder) Fields() map[string]interface{} {
...
+// Adds to an event that happen after it has been sent will return without
+// having any effect.
+func (e *Event) AddField(key string, val interface{}) {
+	e.sendLock.Lock()
+	if e.sent == true {
+		e.sendLock.Unlock()
+		return
+	}
+	e.sendLock.Unlock()
+	e.fieldHolder.AddField(key, val)
```
[the remark this answers is not in the document]
Comment (maplebed) [reply libhoney.go]: eep! yup, thanks.

File: libhoney.go [hunk 18]
```diff
@@ -592,11 +731,15 @@ func (f *fieldHolder) Fields() map[string]interface{} {
 // fields are specified in neither Config nor the Event, Send will return an
 // error.  Required fields are APIHost, WriteKey, and Dataset. Values specified
 // in an Event override Config.
+//
+// Once you Send an event, any addition calls to add data to that event will
+// return without doing anything. Once the event is sent, it becomes immutable.
 func (e *Event) Send() error {
+	dc.ensureLogger()
```
[the remark this answers is not in the document]
Comment (maplebed) [reply libhoney.go]: 👍

File: libhoney.go [hunk 19]
```diff
@@ -580,6 +668,57 @@ func (f *fieldHolder) Fields() map[string]interface{} {
...
+
+// Add adds a complex data type to the event on which it's called.
+// For structs, it adds each exported field. For maps, it adds each key/value.
+// Add will error on all other types.
+//
+// Adds to an event that happen after it has been sent will return without
+// having any effect.
+func (e *Event) Add(data interface{}) error {
+	e.sendLock.Lock()
+	if e.sent == true {
```
[the remark this answers is not in the document]
Comment (maplebed) [reply libhoney.go]: done.

File: libhoney.go [hunk 20]
```diff
@@ -580,6 +668,57 @@ func (f *fieldHolder) Fields() map[string]interface{} {
...
+
+// AddFunc takes a function and runs it repeatedly, adding the return values
+// as fields.
+// The function should return error when it has exhausted its values
+//
+// Adds to an event that happen after it has been sent will return without
+// having any effect.
+func (e *Event) AddFunc(fn func() (string, interface{}, error)) error {
+	e.sendLock.Lock()
+	if e.sent == true {
```
[the remark this answers is not in the document]
Comment (maplebed) [reply libhoney.go]: done

File: transmission/log.go [hunk 21]
```diff
@@ -0,0 +1,13 @@
+package transmission
+
```
[the remark this answers is not in the document]
Comment (maplebed) [reply transmission/log.go]: 👍

File: libhoney.go [hunk 22]
```diff
@@ -580,6 +668,57 @@ func (f *fieldHolder) Fields() map[string]interface{} {
...
+
+// AddField adds an individual metric to the event on which it is called. Note
+// that if you add a value that cannot be serialized to JSON (eg a function or
+// channel), the event will fail to send.
+//
+// Adds to an event that happen after it has been sent will return without
+// having any effect.
+func (e *Event) AddField(key string, val interface{}) {
+	e.sendLock.Lock()
+	if e.sent == true {
```
[the remark this answers is not in the document]
Comment (maplebed) [reply libhoney.go]: I thought I could get away with it but my logic was broken. fixed.

**Comment by ianwilkes** [comment]
> If the uninitialized path is _only_ meant for a noop client, why not a private `initialized` field in `Client` that is only set to `true` in the constructor. Then we can get rid of (i think?) all the `ensure*` calls because we can guarantee that a client is either fully initialized or not initialized at all. the nil client checks could early return on it too, `if client == nil || !client.initialized`

This idea is not terrible. I'm sorry this has led to so much back and forth.

File: client.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (toshok) [reply client.go]: It just makes initialization more consistent - right now there are 3 different patterns for `conf` fallback to initialize different fields in the `Client` struct.

The behavior you're trying to avoid (no `ensure*` call leading to `sync.Once` collision later) is also how the `c.transmission` initialization below acts :/

How about both?

1. change the initialization pattern (for everything but `.builder`, since it requires the backref) to be consistent:
  * Check `conf` setting for nil, fallback by putting the one we want in `conf`
  * `Client` init solely out of `conf`
2. post `Client` init, call c.transmission.Start() so we can early error, then call both `ensureLogger()` and `ensureTransmission()` to use up their `sync.Once`'s before returning the client.

**Comment by maplebed** [comment]
I like the idea of setting an initialized flag in the client and then stripping out the ensures, but I'd like to save that for a future diff. It can be 1.9.1 because everybody loves a point release!
